### ISSUE TITLE
GET on ciscoSpark.internal.encryption.kms.listAuthorizations for resource with large number of authorization fails

### ISSUE THREAD

**Issue by kscreen** [ciscospark/spark-js-sdk#776]
We see an issue when trying to listAuthorizations for a resource with a large number of Authorizations (>100).
We see a loop of errors that start with:
kms: request payload { body:
{ method: 'retrieve',
uri: 'https://encryption-a.wbx2.com/encryption/api/v1/resources/7acfa6a3-6916-4a04-b03e-e16dca4d0ba1/authorizations',
requestId: '01e6a9b3-111e-4697-b7b5-72d5f0824819',
client:
{ clientId: 'https://wdm-a.wbx2.com/wdm/api/v1/devices/338bfb9a-af61-4377-a347-b98354cee1b7',
credential:
{ userId: '76b60da4-0561-482e-80e4-d0cd0ee27f1e',
bearer: 'TOKEN' } } },
requestId: '01e6a9b3-111e-4697-b7b5-72d5f0824819',
uri: 'https://encryption-a.wbx2.com/encryption/api/v1/resources/7acfa6a3-6916-4a04-b03e-e16dca4d0ba1/authorizations',
method: 'retrieve' }
kms: retrieving KMS cluster
kms: batched-request-length 1
**kms: request timed out; request id: 01e6a9b3-111e-4697-b7b5-72d5f0824819; timeout: 6000**
(Note: timeout value varies, I have seen 6000, 12000, and 24000 so far)
**kms: request error KmsError: [object Object]**
KMS_REQUEST_ID: 01e6a9b3-111e-4697-b7b5-72d5f0824819
at KmsError.ExtendableBuiltin (/node_modules/@ciscospark/common/dist/exception.js:50:44)
at KmsError.Exception (/node_modules/@ciscospark/common/dist/exception.js:97:142)
at new KmsError (/node_modules/@ciscospark/internal-plugin-encryption/dist/kms-errors.js:54:123)
at /node_modules/@ciscospark/internal-plugin-encryption/dist/kms-batcher.js:161:20
at process._tickCallback (internal/process/next_tick.js:109:7)
**(node:21199) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 433): KmsError: [object Object]**
KMS_REQUEST_ID: 01e6a9b3-111e-4697-b7b5-72d5f0824819

Errors like this repeat forever.  I have left it running for >1hr and it is still looping.

**Comment by ianwremmel** [comment]
Is this the same as #658?
Is this the same as #774?
Is this the same as #748?

**Comment by ianwremmel** [comment]
What does the unencrypted request payload look like? How do you know how many authorizations the resource has? Have you confirmed with the server folks that KMS is responding in a timely fashion?

There are *a lot* of cases where, rather than supplying an error back to the client, the client simply receives no response. The only steps the client (i.e. the sdk) can take are to retry the request. Once we've retried with a timeout up to 32000ms, the only thing we can assume is that our ecdh token is somehow invalid and therefore renegotiate ecdh and start the process over.

**Comment by kscreen** [comment]
This was opened because #749 was closed and you said to open a new issue if still seeing it.

**Comment by kscreen** [comment]
You are right, it is the same as #748.  I didn't realize that one was still open.

In answer to your question, we know how many authorizations the resource has because we were doing some testing that was adding authorizations and then getting the list.  Once the list got up to 110, it stopped returning and started endlessly looping.
No we have not confirmed anything with KMS server developers.

**Comment by ianwremmel** [comment]
Ok, yea, it would be really helpful to get input from the server folks on this. Is it a particular resource that has 110 authorizations? Or is that that once any given resource hits 110 authorizations, things go wrong?

**Comment by kscreen** [comment]
We have only seen this on one resource which has 110 authorizations, but we have no reason to believe that this wouldn't be a problem for any other resource which has this many authorizations.
Can you create a resource with a lot of authorizations and try it?

**Comment by kscreen** [comment]
Issue #749 was originally created not just to track this issue of not being able to get a list of authorizations on a resource with >100 authorizations, but also because the error returned is not helpful in telling what is wrong.  If you look at the details of what is written in that issue, we hacked the code a little to get back a better error, which provided more information.  It looks like the code has changed somewhat since we did that, but it is still an area that should be improved as we cannot get either results or an informative error.
